iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergAcidUtil.java

```diff
@@ -0,0 +1,145 @@
+    DELETE_FILE_READ_META_COLS.put(MetadataColumns.FILE_PATH, 2);
+    DELETE_FILE_READ_META_COLS.put(MetadataColumns.ROW_POSITION, 3);
+  }
+
+  private static final Types.NestedField PARTITION_HASH_META_COL = Types.NestedField.required(
+      MetadataColumns.PARTITION_COLUMN_ID, MetadataColumns.PARTITION_COLUMN_NAME, Types.LongType.get());
+  private static final Map<Types.NestedField, Integer> DELETE_SERDE_META_COLS = Maps.newLinkedHashMap();
+
+  static {
+    DELETE_SERDE_META_COLS.put(MetadataColumns.SPEC_ID, 0);
+    DELETE_SERDE_META_COLS.put(PARTITION_HASH_META_COL, 1);
+    DELETE_SERDE_META_COLS.put(MetadataColumns.FILE_PATH, 2);
+    DELETE_SERDE_META_COLS.put(MetadataColumns.ROW_POSITION, 3);
+  }
+
+  private static final long RECORD_CACHE_EXPIRE_MILLI = 10 * 60 * 1000;
+  private static final long RECORD_CACHE_MAX_SIZE = 1000;
+  private static final Cache<Schema, GenericRecord> RECORD_CACHE = Caffeine.newBuilder()
+      .expireAfterAccess(RECORD_CACHE_EXPIRE_MILLI, TimeUnit.MILLISECONDS)
+      .maximumSize(RECORD_CACHE_MAX_SIZE)
+      .build();
+
+  /**
+   * @param dataCols The columns of the original file read schema
+   * @param table The table object - it is used for populating the partition struct meta column
+   * @return The schema for reading files, extended with metadata columns needed for deletes
+   */
+  public static Schema createFileReadSchemaForDelete(List<Types.NestedField> dataCols, Table table) {
+    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + DELETE_FILE_READ_META_COLS.size());
+    DELETE_FILE_READ_META_COLS.forEach((metaCol, index) -> {
+      if (metaCol == PARTITION_STRUCT_META_COL) {
+        cols.add(MetadataColumns.metadataColumn(table, MetadataColumns.PARTITION_COLUMN_NAME));
+      } else {
+        cols.add(metaCol);
+      }
+    });
+    cols.addAll(dataCols);
+    return new Schema(cols);
+  }
+
+  /**
+   * @param dataCols The columns of the serde projection schema
+   * @return The schema for SerDe operations, extended with metadata columns needed for deletes
+   */
+  public static Schema createSerdeSchemaForDelete(List<Types.NestedField> dataCols) {
+    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + DELETE_SERDE_META_COLS.size());
+    DELETE_SERDE_META_COLS.forEach((metaCol, index) -> cols.add(metaCol));
+    cols.addAll(dataCols);
+    return new Schema(cols);
+  }
+
+  public static PositionDelete<Record> getPositionDelete(Schema schema, Record rec) {
+    PositionDelete<Record> positionDelete = PositionDelete.create();
+    String filePath = rec.get(DELETE_SERDE_META_COLS.get(MetadataColumns.FILE_PATH), String.class);
+    long filePosition = rec.get(DELETE_SERDE_META_COLS.get(MetadataColumns.ROW_POSITION), Long.class);
+
+    int dataOffset = DELETE_SERDE_META_COLS.size(); // position in the rec where the actual row data begins
+    GenericRecord rowData = RECORD_CACHE.get(schema, GenericRecord::create);
```

Comment:
Maybe we just keep a Record object in the caller and use it as a clone, or set the values

Reply:
Sorry for not being clear enough

Reply:
Oh I see. I think the problem with that is the incoming record object has a different schema (because it has the 4 extra fields at the beginning, such as file_path) than what we want to pass into the PositionDelete, so we'd write out those extra fields too into the delete file

Reply:
as discussed, let's cache the GenericRecord template in the `HiveIcebergDeleteWriter`, and pass it into the util to populate